### PR TITLE
Generate version string from git describe

### PR DESCRIPTION
**Pull request by keszybz**
This was more complex to implement than I thought, and there's an overhead in noop compilations. Nevertheless, I think it's still worth it, because it'll make debugging easier. On both extremes of the development pipeline: when running git installed from a development copy, and when systemd is compiled an distributed as a package.

### CODE REVIEW

**Comment by filbranden** [comment]
Need to update CI so it ships Meson 0.47 before we merge this change...

I also think there are others who really care about version of Meson (e.g. backports to CentOS 7?), so maybe good to start with a heads up to all of those before we push this change.

(Also, from a cursory look, not completely clear to me why newer Meson is needed, maybe it would be good to be more explicit about it in the PR description too.)

Cheers,
Filipe

**Comment by keszybz** [comment]
0.47 is needed for `build_always_stale : true`.

Thinking about this some more, maybe we don't want the .pc files to contain this detailed information? Maybe they should just contain `meson.project_version()` as before. This would  actually make the noop build faster (it's about 0.5s without the regeneration of .pc), and would avoid the need to bump meson version. When writing this patch I just replaced all uses of PACKAGE_VERSION, but maybe that's not needed.

**Comment by poettering** [comment]
Let's do this after v240? Or do you want to see this merged earlier?

**Comment by keszybz** [comment]
Yeah, I don't think we should do the meson version bump now. Do we have a label for this?

**Comment by poettering** [comment]
we do. and i just set it

**Comment by keszybz** [comment]
"good-to-merge" implies just that. And "postponed" doesn't imply anything about quality.

File: tools/meson-vcs-tag.sh [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+dir="$1"
+tag="$2"
+fallback="$3"
+file="${4:-}"
+
+if [ -z "$tag" ]; then
+        # Apparently git describe has a bug where it always considers the work-tree
+        # dirty when invoked with --git-dir (even though 'git status' is happy). Work
+        # around this issue by cd-ing to the source directory.
+        tag="$(cd "$dir" && git describe --dirty=+ 2>/dev/null | sed 's/^v//' || echo "$fallback")"
```
Comment (filbranden) [inline tools/meson-vcs-tag.sh:15]: I typically recommend using an explicit `--abbrev=` as part of the `git describe` command line. (I also recommend something like 12 or 14 for abbrev length.)

For instance, I always have something like `git config core.abbrev 12` or `git config core.abbrev 14` set on my workdirs, to keep the length consistent, but others might not... So making sure the build system will always produce consistent naming regardless of local git config would be great.

**Comment by yuwata** [comment]
Below is the current meson version of each distribution.
\====================
Ubuntu bionic (EOL date: April 2023): [0.45.1](https://packages.ubuntu.com/bionic/meson)
Ubuntu cosmic: [0.47.2](https://packages.ubuntu.com/cosmic/meson)
Ubuntu disco: [0.47.2](https://packages.ubuntu.com/disco/meson)

Debian stretch-backports: [0.47.2](https://packages.debian.org/source/stretch-backports/meson)
Debian buster: [0.49.0](https://packages.debian.org/buster/meson)
Debian sid: [0.49.0](https://packages.debian.org/sid/meson)

Fedora 28: [0.47.2](https://koji.fedoraproject.org/koji/packageinfo?packageID=19835)
Fedora 29,rawhide: [0.49.0](https://koji.fedoraproject.org/koji/packageinfo?packageID=19835)

RHEL 7 with EPEL (End of full support: Q4 of 2019): [0.47.2](https://koji.fedoraproject.org/koji/packageinfo?packageID=19835)

Arch: [0.49.0](https://www.archlinux.org/packages/extra/any/meson/)

openSUSE Tumbleweed: [0.48.2](https://software.opensuse.org/package/meson)

File: tools/meson-vcs-tag.sh (hunk 1 above)
Comment (keszybz) [reply tools/meson-vcs-tag.sh:15]: That's a good point.

**Comment by keszybz** [comment]
> 1. Unpacked archives, where there's no git workdir involved.

Yep, this should work. I didn't actually test with a separate archive, but by simulating a missing .git. Our tarballs are identical to git contents, so this should be enough, but of course a full test would be better.

> 2. A git tree unrelated to this one.

Well, I don't think this is outside of the scope. If somebody plays around with our git repo, then they are responsible for using the configuration switch to override the detection.

> Below is the current meson version of each distribution.

So it seems we can update to 0.47 "for free". We don't support 0.45 anyway.

File: tools/meson-vcs-tag.sh [hunk 2]
```diff
@@ -0,0 +1,16 @@
+#!/bin/sh
+
+set -eu
+set -o pipefail
```
Comment (michaelolbrich) [inline tools/meson-vcs-tag.sh:4]:
```
.../systemd/tools/meson-vcs-tag.sh: 4: set: Illegal option -o pipefail
```
This is with `dash` for `/bin/sh`. I think for `pipefail` /bin/bash is needed.

**Comment by michaelolbrich** [comment]
In `src/udev/udev-ctrl.c` in `ctrl_send()`:
```
strcpy(ctrl_msg_wire.version, "udev-" PACKAGE_VERSION);
```
And `ctrl_msg_wire.version` has only space for 16 bytes. So this will overflow with a long `PACKAGE_VERSION`.

**Pull request closed without merging**